python/pyspark/java_gateway.py

```diff
@@ -41,36 +43,33 @@ def launch_gateway():
         submit_args = submit_args if submit_args is not None else ""
         submit_args = shlex.split(submit_args)
         command = [os.path.join(SPARK_HOME, script)] + submit_args + ["pyspark-shell"]
+
+        # Start a socket that will be used by PythonGatewayServer to communicate its port to us
+        callback_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        callback_socket.bind(('127.0.0.1', 0))
+        callback_socket.listen(1)
+        callback_host, callback_port = callback_socket.getsockname()
+        env = dict(os.environ)
+        env['PYSPARK_DRIVER_CALLBACK_HOST'] = callback_host
```

Comment:
Admittedly, the introduction of new internal environment variables isn't great, but this let me avoid some pain in figuring out how to pipe a hostname / port pair from the Python driver through `spark-submit` all the way to the `PythonGatewayServer`.  If there's a strong feeling that we shouldn't do this, I don't mind figuring out how to implement SparkSubmit-argument-based plumbing instead, but I decided to not do that for the initial cut at fixing this issue.